### PR TITLE
[core][state] Proper report of failure when job finishes

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?
This PR handles cases when a job finishes but tasks still running should be marked as failed. 

- It adds a handler function `OnJobFinished` as a job finish listener in the GcsJobManager, so when a job is marked as finished, the `OnJobFinished` will be called to mark any non-terminated tasks as failed

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests: https://buildkite.com/ray-project/release-tests-pr/builds/25796
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_task_events.py [hunk 1]
```diff
@@ -118,3 +120,140 @@ def verify():
         timeout=10,
         retry_interval_ms=500,
     )
+
+
+def test_fault_tolerance_parents_finish(shutdown_only):
```
Comment (rkooo567) [inline python/ray/tests/test_task_events.py:125]: can you add 2 more cases?

1. Test with actors (nested actors) and actor tasks.
2. Test a complicated e2e case. One idea;
- Start a big tree that has both tasks/actors/actor tasks (that has a height that's higher than 4). 
- Randomly fail one or two of them and make sure all children are marked as failure properly
- Loop these processes 2~3 times.

File: src/ray/common/ray_config_def.h [hunk 2]
```diff
@@ -463,6 +463,11 @@ RAY_CONFIG(int64_t, task_events_max_num_task_events_in_buffer, 10000)
 /// Setting the value to -1 allows unlimited profile events to be sent.
 RAY_CONFIG(int64_t, task_events_max_num_profile_events_for_task, 100)
 
+/// The delay in ms that GCS should mark any running tasks from a job as failed.
+/// Setting this value too smaller might result in some finished tasks marked as failed by
+/// GCS.
+RAY_CONFIG(uint64_t, gcs_mark_task_failed_on_job_done_delay_ms, /*  3 secs */ 1000 * 3)
```
Comment (rkooo567) [inline src/ray/common/ray_config_def.h:469]: Let's have a bit longer delay like 10 seconds, so we can have more time to wait for the report

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 3]
```diff
@@ -663,9 +663,11 @@ void GcsServer::InstallEventListeners() {
       });
 
   // Install job event listeners.
-  gcs_job_manager_->AddJobFinishedListener([this](std::shared_ptr<JobID> job_id) {
-    gcs_actor_manager_->OnJobFinished(*job_id);
-    gcs_placement_group_manager_->CleanPlacementGroupIfNeededWhenJobDead(*job_id);
+  gcs_job_manager_->AddJobFinishedListener([this](const rpc::JobTableData &job_data) {
+    auto job_id = JobID::FromBinary(job_data.job_id());
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_server.cc:667]:
```suggestion
    const auto job_id = JobID::FromBinary(job_data.job_id());
```

File: src/ray/gcs/gcs_server/test/gcs_task_manager_test.cc [hunk 4]
```diff
@@ -441,6 +455,66 @@ TEST_F(GcsTaskManagerTest, TestGetTaskEventsByJob) {
                      reply_job2.mutable_events_by_task());
 }
 
+TEST_F(GcsTaskManagerTest, TestIntermediateTerminatedChild) {
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/test/gcs_task_manager_test.cc:458]: can you add an unit test for the job failure? I believe it is not here?

File: src/ray/gcs/gcs_server/gcs_task_manager.cc [hunk 5]
```diff
@@ -418,5 +492,22 @@ void GcsTaskManager::RecordMetrics() {
       task_event_storage_->GetTaskEventsBytes());
 }
 
+void GcsTaskManager::OnJobFinished(const JobID &job_id, int64_t job_finish_time_ms) {
+  RAY_LOG(DEBUG) << "Marking all running tasks of job " << job_id.Hex() << " as failed.";
+  timer_.expires_from_now(boost::posix_time::milliseconds(
+      RayConfig::instance().gcs_mark_task_failed_on_job_done_delay_ms()));
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_task_manager.cc:498]: Is there any way to know if the job has failed / succeeded? I think we the job has succeeded, it is not necessary to mark children tasks as failed? (they should be eventually reported?)

File: src/ray/gcs/gcs_server/test/gcs_task_manager_test.cc [hunk 6]
```diff
@@ -441,6 +455,66 @@ TEST_F(GcsTaskManagerTest, TestGetTaskEventsByJob) {
...
+  SyncAddTaskEvent({D}, {{rpc::TaskStatus::FAILED, 2}}, C);
+
+  // C running at 3
+  SyncAddTaskEvent({C}, {{rpc::TaskStatus::RUNNING, 3}}, B);
+
+  // B finishes at 4
+  SyncAddTaskEvent({B}, {{rpc::TaskStatus::FINISHED, 4}}, A);
+
+  // A fails at 5
+  SyncAddTaskEvent({A}, {{rpc::TaskStatus::FAILED, 5}});
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/test/gcs_task_manager_test.cc:488]: can you also add a test case where 

1. parent task failure reported
2. child task reported (as non failure state)
3. Make sure both parent & child still fails

File: src/ray/gcs/gcs_server/gcs_task_manager.cc [hunk 7]
```diff
@@ -124,50 +173,75 @@ absl::optional<int64_t> GcsTaskManager::GcsTaskManagerStorage::GetTaskStatusUpda
...
   }
-  auto &task_event = GetTaskEvent(*latest_task_attempt);
-  task_event.mutable_state_updates()->set_failed_ts(failed_ts);
-  task_event.mutable_state_updates()->clear_finished_ts();
+  MarkTaskAttemptFailed(*latest_task_attempt, failed_ts);
 }
 
 void GcsTaskManager::GcsTaskManagerStorage::MarkTaskTreeFailedIfNeeded(
     const TaskID &task_id, const TaskID &parent_task_id) {
+  absl::optional<int64_t> ancestor_failed_ts = absl::nullopt;
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_task_manager.cc:203]: I found this logic a bit hard to follow. Do you see any con if we use recursion here? E.g.,

```
def FailTaskTree(task_id):
    children = get_children(task_id)
    task = get_task_attempt(task_id)

    if (task.state == FAILED):
        # meaning all the children has failed
        return

    task.state = Failed
    for child in children:
        FailTaskTree(child.task_id)
```
?

File: src/ray/common/ray_config_def.h (hunk 2 above)
Comment (rickyyx) [reply src/ray/common/ray_config_def.h:469]: The wait time for the report should be pretty quick IIUC tho? As soon as the driver exits, the raylet will clean up any workers owned by the driver (any running tasks from the job will be killed) and GCS will have the job finished state reported?

File: src/ray/gcs/gcs_server/gcs_task_manager.cc (hunk 5 above)
Comment (rickyyx) [reply src/ray/gcs/gcs_server/gcs_task_manager.cc:498]: if a job succeeds and finishes, any running tasks will be killed, any finished tasks should have their core workers shutdown properly (which will report any task events before exit).

Shouldn't we marked any running tasks failed since they will be killed? E.g. a `time.sleep` that's never awaited and the job finishes.

File: src/ray/gcs/gcs_server/test/gcs_task_manager_test.cc (hunk 6 above)
Comment (rickyyx) [reply src/ray/gcs/gcs_server/test/gcs_task_manager_test.cc:488]: Yeah, I believe this is covered in the `TestFailedParentShouldFailGrandChildren` below?

File: src/ray/gcs/gcs_server/gcs_task_manager.cc (hunk 5 above)
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_task_manager.cc:494]: Do you happen to know where's the logic where raylet kills workers upon the driver exits? I thought this should do the proper cleanup (graceful). 

But it makes sense to do for both cases

File: src/ray/gcs/gcs_server/gcs_task_manager.cc (hunk 7 above)
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_task_manager.cc:203]: > As for the pseudocode you have above, however, would still override a finished child task's state to fail?

Ah I think the exit state of recursion could be 

```
if (task.state == FAILED || task.state == FINISHED):
```
> Also, the above routine will still need to check the parent's failure. So the root is actually different from the children, the root being a failure implies we need to mark potential children as failure, while the children being failure means we could stop at the child.

Hmm is it true though? I assume

1. If the parent has already failed, this child should've been marked as failed. So it will reach to `if (task.state == FAILED || task.state == FINISHED):` and stop recursion. 
2. If the parent hasn't failed yet, it will mark all children as failed. Whe the parent failed, it will check the current node and see "state == FINISHED or FAILED", meaning recursion will stop there

> If it's hard to follow, I should probably do a better job documenting it. What are the parts that are not clear?

I just feel like there's a lot of logic compared to the recursion solution. Mainly I don't 100% understand why we need to check the ancestor timestamp (as I mentioned above ^), but I might be missing something.

File: src/ray/common/ray_config_def.h (hunk 2 above)
Comment (rkooo567) [reply src/ray/common/ray_config_def.h:469]: I think it is safer having a bit longer delay since 1 worker can have lots of tasks buffered, and it takes time to report all. 2. GCS could have been delayed. So I think 3 seconds are pretty short. I originally thought even 30 seconds, but it might be too long lol... I think 10 seconds are not too long and kind of reasonable delay

File: src/ray/gcs/gcs_server/gcs_task_manager.cc (hunk 7 above)
Comment (rickyyx) [reply src/ray/gcs/gcs_server/gcs_task_manager.cc:203]: Even with `task.state ==FINISHED` check,  the recursion will stop incorrectly if there is an intermediate task that finishes. 
E.g. 
`A -> B -> C`
1. All of them running 
2. B finishes (while it is not the most normal use cases but could happen) 
3. A fails

We should mark C as failed in this case? But w/o additional info like ancestor failed ts, there's no way just from failed/finished timestamp to reach C.

File: src/ray/common/ray_config_def.h (hunk 2 above)
Comment (rickyyx) [reply src/ray/common/ray_config_def.h:469]: I see, I thought a worker taking a few seconds to report is more of an edge case while most of the time, all workers should have been shutdown pretty quickly when the job is over, which is more of the common case that the default delay value should cater to. 

But I am open to keep this longer like 10 secs since it's tunable.

File: src/ray/gcs/gcs_server/gcs_task_manager.cc (hunk 7 above)
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_task_manager.cc:203]: Hmm interestingly when I tried your scenario in real ray, C is still running (even after A fails). 

```python3
In [3]: @ray.remote
   ...: def f():
   ...:     pass
   ...: 

In [4]: @ray.remote
   ...: def f():
   ...:     import time
   ...:     time.sleep(300)
   ...: 

In [5]: @ray.remote
   ...: def child():
   ...:     a = f.remote()
   ...: 

In [6]: @ray.remote
   ...: def parent():
   ...:     ray.get(child.remote())
   ...:     import time
   ...:     time.sleep(300)
   ...: 

In [7]: p = parent.remote()
```
this is what I tried (and I manually killed parent ps)

And more interestingly, even after I killed the job, f still keeps running...

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_task_manager.cc:203]: (lol this is why we thought this approach is a bit sketchy... we don't know the exact high level behavior) 

Maybe for these, we should write an integration test to verify if the hypothesis & real scenario is the same. It looks like the grandparent failure doesn't affect the grandchild's failure if the parent didn't fail

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_task_manager.cc:203]: (which also means we probably don't need to handle this ancestor failure? )

File: src/ray/common/ray_config_def.h (hunk 2 above)
Comment (rkooo567) [reply src/ray/common/ray_config_def.h:469]: Let's try 15 seconds (enough delay)!

File: src/ray/gcs/gcs_server/gcs_task_manager.cc (hunk 7 above)
Comment (rickyyx) [reply src/ray/gcs/gcs_server/gcs_task_manager.cc:203]: Ah - great find! I took it for granted on the grandparent - parent - child failure model...

Seems the child will become an orphan if the parent has already finished.  Not sure if it's the most intuitive semantics but yeah, I agree we should probably not handling ancestor failure.

File: src/ray/gcs/gcs_server/gcs_task_manager.cc (hunk 5 above)
Comment (rickyyx) [reply src/ray/gcs/gcs_server/gcs_task_manager.cc:494]: Looks like a worker gets killed if the owning worker dies here: 
https://github.com/ray-project/ray/blob/302a7e5671b51530c08944a505b145af81cdebc7/src/ray/raylet/node_manager.cc#L1074-L1077

File: python/ray/tests/test_task_events.py [hunk 8]
```diff
@@ -118,3 +121,298 @@ def verify():
...
+
+
+def test_fault_tolerance_actor_tasks_failed(shutdown_only):
+    ray.init(_system_config=_SYSTEM_CONFIG)
+    # Test actor tasks
+    with pytest.raises(ray.exceptions.RayTaskError):
+        a = Actor.remote()
+        ray.get(a.fail_parent.remote())
+
+    def verify():
```
Comment (rkooo567) [inline python/ray/tests/test_task_events.py:243]: In the test, is it possible to make sure the result of list_tasks and the real is the same?

I think it will be useful in case we change the behavior in the future (e.g., we allow cascading failures from all the way down to the tree)

File: src/ray/gcs/gcs_server/gcs_task_manager.cc [hunk 9]
```diff
@@ -124,15 +140,29 @@ absl::optional<int64_t> GcsTaskManager::GcsTaskManagerStorage::GetTaskStatusUpda
              : absl::nullopt;
 }
 
+void GcsTaskManager::GcsTaskManagerStorage::MarkTasksFailed(const JobID &job_id,
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_task_manager.cc:143]: Just to make sure I don't miss this; We already have logic to clean children state when the parent fails right? that's why the relevant code has disappeared.

I remember we had N^2 issues as well there. There's no need to resolve in this case?

**Review by rkooo567 — APPROVED** [review]
LGTM. Two last requests!

File: python/ray/tests/test_task_events.py (hunk 8 above)
Comment (rickyyx) [reply python/ray/tests/test_task_events.py:243]: Yeah, this is possible. Let's do this in a follow up PR though? We should probably add some tests that assert the current behavior somewhere.

Comment (rkooo567) [reply python/ray/tests/test_task_events.py:243]: that works for me

**Comment by rkooo567** [comment]
Btw note; the PR description may be wrong. We only mark children failed (not grandchildren)

**Comment by rkooo567** [comment]
Let's also close the P0 issue!

File: python/ray/tests/test_task_events.py (hunk 8 above)
Comment (rickyyx) [reply python/ray/tests/test_task_events.py:243]: Tracked ehre https://github.com/ray-project/ray/issues/31875

**Comment by rickyyx** [comment]
Updated PR description and title.

**Comment by rkooo567** [comment]
Let's also make sure to not forget the follow up test!

**Comment by rickyyx** [comment]
Yep here: https://github.com/ray-project/ray/issues/31875
